v23.1/transaction-retry-error-example.md

```diff
@@ -0,0 +1,118 @@
+
+## Client-side retry handling example
+
+{% include {{page.version.version}}/misc/client-side-intervention-example.md %}
+
+## Testing transaction retry logic
+
+To test your application's transaction retry logic, use the [`inject_retry_errors_enabled` session variable](set-vars.html#supported-variables). When `inject_retry_errors_enabled` is set to `true`, any statement (with the exception of [`SET` statements](set-vars.html)) executed in the session inside of an explicit transaction will return a [transaction retry error](transaction-retry-error-reference.html) with the message ```restart transaction: TransactionRetryWithProtoRefreshError: injected by `inject_retry_errors_enabled` session variable```.
+
+If the client retries the transaction using the special [`cockroach_restart` `SAVEPOINT` name](savepoint.html#savepoints-for-client-side-transaction-retries), after the 3rd retry, the transaction will proceed as normal. Otherwise, the errors will continue until the client issues a `SET inject_retry_errors_enabled=false` statement.
+
+For example, suppose you've written a wrapper function with some retry logic that you want to use to execute statements across a [`psycopg2`](https://www.psycopg.org/) connection:
+
+~~~ python
+def run_transaction(conn, op, max_retries=3):    
+    """
+    Execute the operation *op(conn)* retrying serialization failures.
+
+    If the database returns an error asking to retry the transaction, retry it
+    *max_retries* times before giving up (and propagate it).
+    """
+    # leaving this block the transaction will commit or rollback
+    # (if leaving with an exception)
+    with conn:
+        for retry in range(1, max_retries + 1):
+            try:
+                op(conn)
+                # If we reach this point, we were able to commit, so we break
+                # from the retry loop.
+                return
+
+            except SerializationFailure as e:
+                # This is a retry error, so we roll back the current
+                # transaction and sleep for a bit before retrying. The
+                # sleep time increases for each failed transaction.
+                logging.debug("got error: %s", e)
+                conn.rollback()
+                sleep_ms = (2 ** retry) * 0.1 * (random.random() + 0.5)
+                logging.debug("Sleeping %s seconds", sleep_ms)
+                time.sleep(sleep_ms)
+
+            except psycopg2.Error as e:
+                logging.debug("got error: %s", e)
+                raise e
+
+        raise ValueError(f"Transaction did not succeed after {max_retries} retries")
+~~~
+
+`run_transaction` takes a SQL-executing function `op(conn)` and attempts to run the function, retrying on serialization failures (exposed in pscyopg2 as the [`SerializationFailure` exception class](https://www.psycopg.org/docs/errors.html#sqlstate-exception-classes)) with exponential backoff, until reaching a maximum number of tries.
```

Comment:
This explanation of the function makes more sense to me above the code block? Maybe like "suppose you've written a wrapper function `run_transaction`, which takes a ..."